docs/API/config_pydantic.md
# 

## **Config**

The Config class is a [`pydantic.BaseModel`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel) that defines model parameters. The parameters are categorized into four groups, each catering to different aspects of the super-resolution (SR) evaluation process:

### **Global Parameters:**

These parameters control the overall behavior of the SR assessment:

- **device:** Selects the computation device for inference, defaulting to `cpu`. `cuda` is also supported.

- **agg_method:** Determines the granularity for distance metric application — pixel-wise by default ("pixel"), with "patch" and "image" level aggregations options. If "patch" is selected, the metric is computed on patches of size `patch_size` and then interpolated to the original image size using bicubic interpolation with anti-aliasing. If "image" is selected, the metric is computed on the entire image and output as a single value. If "pixel" is selected, the metric is computed on each pixel and output as a tensor of the same size as the input image.

- **patch_size:** Relevant when `agg_method` is "patch", this defines the patch size for distance metric computation.

- **border_mask:** Excludes image borders in metric calculations, ignoring the outer 16 pixels by default.

- **rgb_bands:** Necessary for certain metrics like LPIPS, CLIP and plot generation, assuming a default order of Red-Green-Blue ([0, 1, 2]).

- **harm_apply_spectral:** Applies histogram matching before correctness metrics. This is enabled by default.

- **harm_apply_spatial:** Activates spatial alignment before correctness metrics. This is enabled by default.

### **Spatial Parameters:**

These parameters set the spatial alignment pre-processing:

- **spatial_method:** The default method for spatial alignment. By default, pcc (Phase Correlation Coefficient) is used. However, other methods like "ecc" (Enhanced Correlation Coefficient) and "lgm" (SuperPoint + LightGlue) are also available. ecc is more robust to noise and usually more precise than pcc, but it is slower. lgm is slower than ecc and pcc, but it is more robust to large translations.

- **spatial_threshold_distance**: The maximum permissible translation distance for spatial alignment, set to 5 pixels by default. If the translation distance is greater than this threshold, the spatial alignment is skipped.

- **spatial_max_num_keypoints:** Only relevant when the spatial method is "lgm". This parameter caps the number of keypoints for feature matching at 500 by default.


### **Spectral and reflectance Parameters:**

These parameters are specific to spectral analysis and alignment:

- **reflectance_distance:** The default method for reflectance distance estimation. Reflectance distance measure how the SR image affects the reflectance of the original image. By default, "l1" is used. See section [Metrics](docs/Metrics/consistency.md) for more details and other options.

- **spectral_distance:** The default method for spectral distance calculation.  The spectral distance measure how the SR image affects the spectral signature of the original image. By default, "sam" is used. See section [Metrics](docs/Metrics/consistency.md) for more details and other options.

### **Synthesis parameters**

- **synthesis_distance:** Specifies the distance metric between SR and LR after harmonization. By default, "l1" is used. See section [Metrics](docs/Metrics/consistency.md) for more details and other options.

### **Correctness parameters**

- **correctness_distance:** Specifies the distance metric between harmonized SR, HR and LR. By default, "l1" is used. See section [Metrics](docs/Metrics/consistency.md) for more details and other options.


Below is an illustrative example of how to instantiate `opensr-test` with user-defined parameters:

```python
import torch
import opensr_test

# Define the Low Resolution (LR), High Resolution (HR), and Super-Resolved (SR) images.
lr = torch.rand(4, 64, 64)
hr = torch.rand(4, 256, 256)
sr = torch.rand(4, 256, 256)

# Initialize the Metrics object with custom parameters.
config = opensr_test.Config(
    device="cuda",
    spatial_features="ecc",
    harm_apply_spatial=True,
    harm_apply_spectral=False
)
metrics = opensr_test.Metrics(config)

# Compute the metrics based on the provided images.
metrics.compute(lr=lr, sr=sr, hr=hr)
```

docs/API/model_parameters.md
# 

## Model Parameters

`opensr-test` utilizes [pydantic](https://docs.pydantic.dev/latest/) for defining model parameters, ensuring type safety and ease of configuration. The parameters are categorized into three groups, each catering to different aspects of the super-resolution (SR) evaluation process:

### Global Parameters:

These parameters control the overall behavior of the SR assessment:

- **device:** Selects the computation device for inference, defaulting to `cpu`. `cuda` is also supported for GPU acceleration.
- **distance_method:** Specifies the distance metric between SR, LR, and HR images. Defaults to "L1", with support for other metrics like KL, L2, PBIAS, RMSE, 1/PSNR, SAD, CROSSMTF, CLIP and LPIPS. See section [Metrics](docs/Metrics/consistency.md) for more details.
- **agg_method:** Determines the granularity for distance metric application — pixel-wise by default ("pixel"), with "patch" and "image" level aggregations options.
- **patch_size:** Relevant when `agg_method` is "patch", this defines the patch size for distance metric computation.
- **border_mask:** Excludes image borders in metric calculations, ignoring the outer 16 pixels by default.
- **rgb_bands:** Necessary for certain metrics like LPIPS and plot generation, assuming a default order of Red-Green-Blue ([0, 1, 2]).
- **harm_apply_spectral:** Applies histogram matching before distance metric calculation, enabled by default.
- **harm_apply_spatial:** Activates spatial alignment before distance metric calculation, enabled by default.

### Spatial Parameters:

These parameters fine-tune the spatial feature extraction and matching process:

- **spatial_features:** The default feature extractor is "superpoint", with "disk" also available.
- **spatial_matcher:** Sets the matcher for feature points, defaulting to "lightglue".
- **spatial_max_num_keypoints:** Caps the number of keypoints for feature matching at 1000 by default.
- **spatial_threshold_distance:** The maximum permissible distance between keypoints for a match, set to 5 pixels.
- **spatial_threshold_npoints:** The minimum required keypoints for alignment, set to 5. Spatial alignment is skipped if keypoints fall below this threshold.

### Spectral Parameters:

These parameters are specific to spectral analysis and alignment:

- **reflectance_method:** The default method for reflectance estimation is "l1", with "kl", "l1", "l2", and "pbias" as alternatives.

- **spectral_method:** For spectral alignment, "sad" is the default method.

Below is an illustrative example of how to instantiate `opensr-test` with user-defined parameters:

```python
import torch
import opensr_test

# Define the Low Resolution (LR), High Resolution (HR), and Super-Resolved (SR) images.
lr = torch.rand(4, 32, 32)
hr = torch.rand(4, 256, 256)
sr = torch.rand(4, 256, 256)

# Initialize the Metrics object with custom parameters.
metrics = opensr_test.Metrics(
    device="cpu",
    spatial_features="disk",
    spatial_max_num_keypoints=400,
    harm_apply_spectral=False,
)

# Setup the evaluation environment with the LR, SR, and HR images.
metrics.setup(lr=lr, sr=sr, hr=hr)

# Compute the metrics based on the provided images.
metrics.compute()

# Generate a summary report of the evaluation.
metrics.summary()
```

NEWS.md
# 0.2.1

- torchvision dependency added to setup.py.

# 0.2.0

- Several bugs have been fixed.


# 0.1.2

- Bug in visualization triplets and quads fixed.



# 0.1.0 Initial release

- First version of the package now available on pypi.


opensr_test/config.py
from typing import Any, List, Optional, Union, Literal

import torch
from pydantic import BaseModel, field_validator, model_validator

DistanceMetrics = Literal[
    "kl", "l1", "l2", "pbias", "psnr", "sad",
    "mtf", "lpips", "clip", "fd"
]

class Config(BaseModel):
    # General parameters
    device: Union[str, Any] = "cpu"
    agg_method: str = "pixel"  # pixel, image, patch
    patch_size: Optional[int] = None
    border_mask: Optional[int] = 16
    rgb_bands: Optional[List[int]] = [0, 1, 2]
    harm_apply_spectral: bool = True
    harm_apply_spatial: bool = True
    clip_model_path: Optional[str] = None


    # Spatial parameters
    spatial_method: Literal["ecc", "pcc", "lgm"] = "pcc"
    spatial_threshold_distance: int = 5
    spatial_max_num_keypoints: int = 500

    # Spectral parameters
    reflectance_distance: DistanceMetrics = "l1"
    spectral_distance: DistanceMetrics = "sad"


    # Synthesis parameters
    synthesis_distance: DistanceMetrics = "l1"

    # Correctness parameters
    correctness_distance: DistanceMetrics = "fd"
    correctness_norm: Literal["softmin", "percent"] = "softmin"
    im_score: Optional[float] = 0.05
    om_score: Optional[float] = 0.05
    ha_score: Optional[float] = 0.05 # be quiet conservative about hallucinations
    correctness_temperature: Optional[float] = 0.25 # be quiet conservative about hallucinations


    # General parameters - validator ----------------------------
    @field_validator("device")
    def check_device(cls, value) -> str:
        return torch.device(value)

    @field_validator("agg_method")
    def check_agg_method(cls, value) -> str:
        valid_methods = ["pixel", "image", "patch"]
        if value not in valid_methods:
            raise ValueError(f"Invalid method. Must be one of {valid_methods}")
        return value

    @field_validator("rgb_bands")
    def check_rgb_bands(cls, value) -> List[int]:
        if len(value) != 3:
            raise ValueError("rgb_bands must have 3 elements.")
        return value

    @field_validator("spatial_max_num_keypoints")
    def check_spatial_max_num_keypoints(cls, value) -> int:
        if value < 0:
            raise ValueError("spatial_max_num_keypoints must be positive.")
        return value

    @field_validator("spatial_threshold_distance")
    def check_spatial_threshold_distance(cls, value) -> int:
        if value < 0:
            raise ValueError("spatial_threshold_distance must be positive.")
        return value


    # Create SRharm - validator ------------------------------------------
    @field_validator("harm_apply_spectral")
    def check_harm_apply_spectral(cls, value) -> bool:
        if not isinstance(value, bool):
            raise ValueError("harm_apply_spectral must be boolean.")
        return value

    @field_validator("harm_apply_spatial")
    def check_harm_apply_spatial(cls, value) -> bool:
        if not isinstance(value, bool):
            raise ValueError("harm_apply_spatial must be boolean.")
        return value


class Consistency(BaseModel):
    reflectance: Any
    spectral: Any
    spatial: Any

class Synthesis(BaseModel):
    distance: Any

class Correctness(BaseModel):
    omission: Any
    improvement: Any
    hallucination: Any
    classification: Any

class Auxiliar(BaseModel):
    sr_harm: Any
    lr_to_hr: Any
    d_ref: Any
    d_im: Any
    d_om: Any

class Results(BaseModel):
    consistency: Consistency
    synthesis: Synthesis
    correctness: Correctness
    auxiliar: Auxiliar

